feat(server): Fix exif data parsing

samip5 opened the pull request:
Moved to the exiftool-vendored library and as that requires the binary on the container, I also edited the Dockerfile for that.

I have checked that this works now, but I don't have images to test with that lack exif data.

samip5 committed:
Trying to get exifdata working with different lib.

samip5 committed:
Got the new library working.

alextran1502 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -144,29 +144,22 @@ export class MetadataExtractionProcessor {
-      });
+      await exiftool.version().then((version: any) => Logger.debug(`We are running ExifTool v${version}`));
+      const exifData = await exiftool.read(asset.originalPath);
+      Logger.log(exifData);
```
We probably want to remove this line, otherwise, it will flood the console

jrasm91 replied:
We could leave it, but use verbose instead.

samip5 replied:
Ah, yeah or at least changed to debug but probably best to just remove?

jrasm91 replied:
Can we remove the version log too? If anything, only do it once at startup (in the constructor, instead of for every asset it processes.

samip5 replied:
Removed both of those entirely.

jrasm91 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -175,53 +168,14 @@ export class MetadataExtractionProcessor {
-          });
-        }
-      } else {
-        await this.assetRepository.save({
-          id: asset.id,
-          createdAt: createdAt.toISOString(),
```
We should still keep this as the original created date is passed in by the client and may be incorrect.

samip5 replied:
Do you mean the whole block of code? Related commen: https://discord.com/channels/979116623879368755/1056815231256514660/1063943667842289857

jrasm91 replied:
In both situations the created date on the asset gets updated, regardless if there was a timezone issue or not.

I think the whole "fix the timezone block" can be removed, but we should still save the new date back on the original asset.

samip5 replied:
Preserved.

jrasm91 reviewed (commented):
What is the benefit of this library vs the current one? It automatically resolves the timezones inconsistencies?

samip5 commented:
> What is the benefit of this library vs the current one? It automatically resolves the timezones inconsistencies?

It would hopefully fix the issues with timezones like #676, possibly others. The library handles automatically the timezone parsing, which I believe would help.

I believe that @alextran1502 was meaning to move to this library anyway as discussed in Discord. https://discord.com/channels/979116623879368755/1056815231256514660/1059905637460484197

samip5 committed:
Addressing PR comments.

samip5 committed:
Removed not used vars and proper place for the eslint disable.

samip5 committed:
Fix time-utils to use the exiftool-vendored lib.

Fixed also one test, as that would be valid.

alextran1502 commented:
The unit test failed, can you please resolve it?

samip5 committed:
Using filename for timestamp as well if possible.

samip5 committed:
Add new tests for time-utils.

samip5 commented:
> The unit test failed, can you please resolve it?

I cannot see which test that is (in Vercel), and my local machine has two tests that are failing. Are we talking about the same ones?

```
 FAIL  apps/immich/src/config/asset-upload.config.spec.ts
  ● Test suite failed to run

    Vendored ExifTool does not exist at /home/sky/Github/immich/server/node_modules/exiftool-vendored.pl/bin/exiftool

    > 1 | import { exiftool } from 'exiftool-vendored';
        | ^
      2 |
      3 | function createTimeUtils() {
      4 |   const checkValidTimestamp = (timestamp: string): boolean => {

      at findExiftool (node_modules/exiftool-vendored/src/ExifTool.ts:135:9)
      at Object.<anonymous> (node_modules/exiftool-vendored/src/ExifTool.ts:138:36)
      at Object.<anonymous> (libs/common/src/utils/time-utils.ts:1:1)
      
 FAIL  apps/immich/src/config/profile-image-upload.config.spec.ts
  ● Test suite failed to run

    Vendored ExifTool does not exist at /home/sky/Github/immich/server/node_modules/exiftool-vendored.pl/bin/exiftool

    > 1 | import { exiftool } from 'exiftool-vendored';
        | ^
      2 |
      3 | function createTimeUtils() {
      4 |   const checkValidTimestamp = (timestamp: string): boolean => {

      at findExiftool (node_modules/exiftool-vendored/src/ExifTool.ts:135:9)
      at Object.<anonymous> (node_modules/exiftool-vendored/src/ExifTool.ts:138:36)
      at Object.<anonymous> (libs/common/src/utils/time-utils.ts:1:1)

```

samip5 committed:
Remember to gracefully terminate the exiftool instance when not needed.

samip5 committed:
eslint ignore...

alextran1502 commented:
Hmm there is currently some problems with the PR. I cannot run the PR successfully, none of the EXIF is registered to the database. I will have to take a look at this again.

alextran1502 commented:
![image](https://user-images.githubusercontent.com/27055614/212526234-d3c0e062-4781-4583-8db8-8d61d4823fd9.png)
 lots of these

samip5 commented:
> ![image](https://user-images.githubusercontent.com/27055614/212526234-d3c0e062-4781-4583-8db8-8d61d4823fd9.png) lots of these

Please make sure that exiftool exists in the container. I'm quite confused why it wouldn't be despite the Dockerfile should be installing it there.

`docker exec -it immich_microservices sh` and `apk add exiftool` and those errors go away.

samip5 committed:
Apperantly Dockerfile changes were not pushed.

samip5 committed:
feat(dockerfile): Tweak the Server Dockerfile

jbaez commented on server/libs/common/src/utils/time-utils.ts:
```diff
@@ -19,30 +19,33 @@ function createTimeUtils() {
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  const getTimestampFromFilename = async (originalPath: string): Promise<string> => {
```
could we remove the `@ts-ignore` (and eslint-disable)? This function would return either a promise that resolves to a string or undefined.

Then we can check if it resolves with a timestamp string in `server/apps/microservices/src/processors/metadata-extraction.processor.ts`  like:
```typescript
     if (!exifData || ! timestampFromFilename) {
        throw new Error(`can not parse exif data from file ${asset.originalPath}`);
      }
```

We can also add a test scenario for when it returns undefined (hypothetical case were the filename has a different naming format that doesn't match the current regex) 😃

samip5 replied:
That sounds like a good idea. However I'm not quite sure as to what kind of test case should be added for that hypothetical case. :D

samip5 committed:
feat(server): getTimestampFromFilename should return string or undefined.

samip5 committed:
feat(server): If we don't have exifData or timestamp from filename, raise an error.

alextran1502 commented:
Hello, I take a look again at the PR and discovered some issues

## Cannot run in development environment.
 
I believe you have to include `exiftool` `perl` to the apk package in Dockerfile in order to have the exif tool run in development environment as indicated in the library's readme
![image](https://user-images.githubusercontent.com/27055614/212563191-bf3f1200-91b2-4c80-82ee-c5f59cac679f.png)

Here is my current Dockerfile that ensure the library can run in development mode

![image](https://user-images.githubusercontent.com/27055614/212563417-f821ab69-e4bd-47de-a268-05e909bf7175.png)

This raises a question: 

How did you test your code?

## Code issue

![image](https://user-images.githubusercontent.com/27055614/212563261-b9a57d51-826c-486e-9267-c644161e4ab7.png)

`exiftool.read()` always return a tag otherwise it will throw an error. So the function `timeUtils.getTimestampFromFilename()` (Currently broken) is undefined everytime so the condition on line 149 will always avaluate `!timestampFromFilename` and throw an error.

samip5 committed:
Apparently test was already right, but my local system disagrees.

samip5 committed:
More utilities for parsing and fix the timestampFromFilename.

It was returning an incorrect date as the regex doesn't seem to be the best for this as files named `IMG_0115.HEIC` will want to get parsed incorrectly due to it.

samip5 committed:
feat(server/docker): Install perl as it seems to be required.

samip5 committed:
feat(server): remember to include exposureTime and focalLength in new exif data.

samip5 commented:
> How did you test your code?

I'm using IntelliJ WebStorm, and I'm running the development docker compose environment that reloads the code on changes.

I have been testing the code with just going to the web interface and uploading photos and checking if they do have exif data visible while also checking the log of microservices container. 
The two tests (when running the server tests) that fail are a result of exiftool not finding a vendored exiftool despite it existing in the path it lists (eg. `/home/sky/Github/immich/server/node_modules/exiftool-vendored.pl/bin/exiftool`). 


# Code issues

I noticed that my regex for getting timestamp from filename is not quite good, as files named `IMG_0115.HEIC` for example like to fail due to it, good ideas on how it should be improved?

I haven't gotten the error for `can not parse exif data` so to me it is working correctly?

alextran1502 commented:
> > How did you test your code?
> 
> I'm using IntelliJ WebStorm, and I'm running the development docker compose environment that reloads the code on changes.
> 
> I have been testing the code with just going to the web interface and uploading photos and checking if they do have exif data visible while also checking the log of microservices container. The two tests (when running the server tests) that fail are a result of exiftool not finding a vendored exiftool despite it existing in the path it lists (eg. `/home/sky/Github/immich/server/node_modules/exiftool-vendored.pl/bin/exiftool`) which I'm really confused about as to why it still deems it to not exist.
> 
> ## Code issues
> I noticed that my regex for getting timestamp from filename is not quite good, as files named `IMG_0115.HEIC` for example like to fail due to it, good ideas on how it should be improved?
> 
> I haven't gotten the error for `can not parse exif data` so to me it is working correctly (or was it just caught on line 220)? Let me do more testing.

Interesting, without adding the additional packages in the Docker file, I am not sure how it worked on your ends but didn't work on my setup 😭 

For testing, I was using my iPhone to upload to the server. Have you tried that route?

alextran1502 commented:
Another thought is that you don't have to use getting timestamp from filename, just grab it directly from the `asset.createdAt` property, since the filename info is built based on that anyway. And different setup would have different storage template setup so the getting the info from that isn't the best route to take.

samip5 commented:
> For testing, I was using my iPhone to upload to the server. Have you tried that route?

I haven't tried that route as my phone doesn't have the iPhone raw photos due to using an Android but used to use an Apple device so for me to be able to test multiple formats, it's simpler to just drag & drop them to the Web UI from my computer.

> Another thought is that you don't have to use getting timestamp from filename, just grab it directly from the `asset.createdAt`

If the filename includes the correct date in the filename, shouldn't we use it rather than rely on `asset.createdAt`? My goal with that is so if it is included, it could be used and thus hopefully end up with proper data than possible via exif.

alextran1502 commented:
> > For testing, I was using my iPhone to upload to the server. Have you tried that route?
> 
> I haven't tried that route as my phone doesn't have the iPhone raw photos due to using an Android but used to use an Apple device so for me to be able to test multiple formats, it's simpler to just drag & drop them to the Web UI from my computer.
> 
> > Another thought is that you don't have to use getting timestamp from filename, just grab it directly from the `asset.createdAt`
> 
> If the filename includes the correct date in the filename, shouldn't we use it rather than rely on `asset.createdAt`? My goal with that is so if it is included, it could be used and thus hopefully end up with proper data than possible via exif.

I am not sure I understand the intention here. The file path is customizable, so you don't know all the possible format to parse the string correctly. The the file path is derived from the `asset.createdAt` regardless. So instead of manually parsing the file path, why don't we use the `asset.createdAt` instead?

samip5 commented:
> > > For testing, I was using my iPhone to upload to the server. Have you tried that route?
> > 
> > 
> > I haven't tried that route as my phone doesn't have the iPhone raw photos due to using an Android but used to use an Apple device so for me to be able to test multiple formats, it's simpler to just drag & drop them to the Web UI from my computer.
> > > Another thought is that you don't have to use getting timestamp from filename, just grab it directly from the `asset.createdAt`
> > 
> > 
> > If the filename includes the correct date in the filename, shouldn't we use it rather than rely on `asset.createdAt`? My goal with that is so if it is included, it could be used and thus hopefully end up with proper data than possible via exif.
> 
> I am not sure I understand the intention here. The file path is customizable, so you don't know all the possible format to parse the string correctly. The the file path is derived from the `asset.createdAt` regardless. So instead of manually parsing the file path, why don't we use the `asset.createdAt` instead?

I mean that if the filename is like `IMG_20220210_456.jpg`, and exif data doesn't have the correct date, I think it should try to parse it and if it fails use the file creation date. Does that make sense?

alextran1502 commented:
> > > > For testing, I was using my iPhone to upload to the server. Have you tried that route?
> 
> > > 
> 
> > > 
> 
> > > I haven't tried that route as my phone doesn't have the iPhone raw photos due to using an Android but used to use an Apple device so for me to be able to test multiple formats, it's simpler to just drag & drop them to the Web UI from my computer.
> 
> > > > Another thought is that you don't have to use getting timestamp from filename, just grab it directly from the `asset.createdAt`
> 
> > > 
> 
> > > 
> 
> > > If the filename includes the correct date in the filename, shouldn't we use it rather than rely on `asset.createdAt`? My goal with that is so if it is included, it could be used and thus hopefully end up with proper data than possible via exif.
> 
> > 
> 
> > I am not sure I understand the intention here. The file path is customizable, so you don't know all the possible format to parse the string correctly. The the file path is derived from the `asset.createdAt` regardless. So instead of manually parsing the file path, why don't we use the `asset.createdAt` instead?
> 
> 
> 
> I mean that if the filename is like `IMG_20220210_456.jpg`, and exif data doesn't have the correct date, I think it should try to parse it and if it fails use the file creation date. Does that make sense?

Ah got it, but what are the chances the file has such name? I assume this is from your use case?

samip5 commented:
> > I mean that if the filename is like `IMG_20220210_456.jpg`, and exif data doesn't have the correct date, I think it should try to parse it and if it fails use the file creation date. Does that make sense?
> 
> Ah got it, but what are the chances the file has such name? I assume this is from your use case?

I believe that it's actually quite likely that it has such a name, at least Android's default Camera app seems to save the files in such format which is why I ended up on that idea. It seem iPhone doesn't as it uses just a generic identifier and not a date.

samip5 commented:
The tests failing seems to be an issue with jest's ability to locate `exiftool-vendored.pl`, but I wonder how that's different from normal node's ability to do the same.

Related: https://github.com/photostructure/exiftool-vendored.js/issues/128

alextran1502 commented:
> I believe that it's actually quite likely that it has such a name, at least Android's default Camera app seems to save the files in such format which is why I ended up on that idea. It seem iPhone doesn't as it uses just a generic identifier and not a date.

I think the format only applies to a certain phone, and different phones might have a different format. I just tested on the Samsung S8, and the file name is something like this. `20220811_085512.jpg` 

I would like to not include edge cases in the code.

samip5 commented:
> > I believe that it's actually quite likely that it has such a name, at least Android's default Camera app seems to save the files in such format which is why I ended up on that idea. It seem iPhone doesn't as it uses just a generic identifier and not a date.
> 
> I think the format only applies to a certain phone, and different phones might have a different format. I just tested on the Samsung S8, and the file name is something like this. `20220811_085512.jpg`
> 
> I would like to not include edge cases in the code.

It would parse that correctly as well, so should we just not include it then?

alextran1502 commented:
> > > I believe that it's actually quite likely that it has such a name, at least Android's default Camera app seems to save the files in such format which is why I ended up on that idea. It seem iPhone doesn't as it uses just a generic identifier and not a date.
> > 
> > 
> > I think the format only applies to a certain phone, and different phones might have a different format. I just tested on the Samsung S8, and the file name is something like this. `20220811_085512.jpg`
> > I would like to not include edge cases in the code.
> 
> It would parse that correctly as well, so should we just not include it then?

That would only parse the date part but not the time part, correct?

samip5 commented:
Correct as I hadn't realized that the part after _ is time.

alextran1502 commented:
After some consideration, I would like to remove such a mechanism because of the differences of filenames across different Android phones. Let's stick with EXIF for the date time information

samip5 committed:
feat(server): Remove the parsing from filename as requested.

samip5 commented:
Okay, I have removed the filename parsing as requested.

Do you have ideas how to fix the tests?

alextran1502 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -144,84 +143,43 @@ export class MetadataExtractionProcessor {
-      });
+      const exifData = await exiftool.read(asset.originalPath);
 
       if (!exifData) {
```
This line will not do anything because you will always have `exifData` if a file doesn't have exif data an error will be thrown. So you might have to do a catch directly in the exif data read method 

```typescript
const exifData = await exiftool.read(asset.originalPath).catch(e => console.log("No exif"));
```

And then resolve the scenario when no EXIF data is available.

samip5 committed:
feat(server): Import exiftool differently in time-utils.

samip5 committed:
feat(server): Error handling when there is no exifData.

alextran1502 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -144,84 +143,42 @@ export class MetadataExtractionProcessor {
-        ihdr: true,
+      const exifData = await exiftool.read(asset.originalPath).catch(e => {
+        this.logger.debug(`The exifData parsing failed due to: ${e}`);
+        throw new Error(`Can not parse exif data from file ${asset.originalPath}`);
```
Perhaps just catch the error and log it; throwing an error will result in the function terminating here, correct?

samip5 replied:
Yeah, just noticed that and also pushed a fix for it.

alextran1502 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -144,84 +143,42 @@ export class MetadataExtractionProcessor {
+        exifDate ? new Date(exifDate.toString()!) : null;
 
+      const createdAt = exifToDate(exifData.DateTimeOriginal ?? exifData.CreateDate ?? asset.createdAt);
+      const modifyDate = exifToDate(exifData.ModifyDate);
```
Should we need to include `asset.modifiedAt` here when the EXIF data is not available?

samip5 replied:
I think that would be beneficial so I have pushed changes to reflect that.

alextran1502 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -144,84 +143,42 @@ export class MetadataExtractionProcessor {
       newExif.imageName = path.parse(fileName).name || null;
       newExif.exifImageHeight = exifData['ExifImageHeight'] || exifData['ImageHeight'] || null;
       newExif.exifImageWidth = exifData['ExifImageWidth'] || exifData['ImageWidth'] || null;
+      newExif.exposureTime = (await timeUtils.parseStringToNumber(exifData['ExposureTime'])) || null;
```
What is the different of using this method vs `toString()`?

samip5 replied:
The difference is typing, as that takes either an `string` or `undefined` and outputs either `number` or `null` which it wasn't happy about as `newExif.exposureTime` needs a number, not a string.

samip5 committed:
feat(server): Fixes for the error handling when there is no exifData.

samip5 committed:
feat(server): Remember to include modifyDate despite no exif.

samip5 committed:
feat(server): Remember to include model of Camera.

samip5 committed:
Merge remote-tracking branch 'origin' into feat/exif

samip5 committed:
feat(server): Fixing up Exiftool usage.

Including proper logging for it, which had to be done in wrapped fashion due to it expecting all the logging levels which NextJS logger doesn't implement.

samip5 commented:
I really don't like the way I had to do logging from the ExifTool as it involved making a wrapper for the NextJS logger, but what do you think?

alextran1502 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -257,6 +261,7 @@ export class MetadataExtractionProcessor {
       }
 
       await this.exifRepository.save(newExif);
+      await exiftool.end();
```
Adding this line will shutdown the exiftool on other processes. In order to make this work, we have to remove this line. Probably due to the way we are initiating the exif-vendor instance. 

I tested this from uploading from the mobile app, continuously for about 200 assets

samip5 replied:
Which commit was the one you tested with? I just a moments ago pushed more changes, like the way we are initializing the instance for it.

alextran1502 replied:
Got it. The recent commit works

alextran1502 commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -73,10 +72,48 @@ export type GeoData = {
   distance: number;
 };
 
+class WrapperLogger {
```
We've already had a logger configured. Please don't create a wrapper.

See example
https://github.com/samip5/immich/blob/92eee11e8d7e9173518be4b52b8e4deea1ac0290/server/apps/immich/src/api-v1/asset/asset.service.ts#L70-L71

samip5 replied:
Unfortunately, without a wrapper:

```
TS2322: Type '() => import("/home/sky/Github/immich/server/node_modules/@nestjs/common/services/logger.service").Logger' is not assignable to type '() => import("/home/sky/Github/immich/server/node_modules/batch-cluster/dist/Logger").Logger'.   Type 'Logger' is missing the following properties from type 'Logger': trace, info
```

The NextJS logger doesn't implement those, so what's the better way to handle that?

alextran1502 replied:
We don't have to use the internal logger. The information from the logic flow would leave us enough info for trouble shooting. 

Sorry for the back and forth, we are both working too fast 😛

samip5 replied:
True we don't have to, but then that would need to imported from some other package, as one cannot use NextJS one.

I thought that it would be better to use the internal one so that it's not out of place in the log output.

Which approach we then should take? Import the logger from somewhere else and not use a wrapper for it?

alextran1502 replied:
Let's not use the wrapper for it.

samip5 replied:
ALright, removed the wrapper and it uses the batch-cluster's default logger.

alextran1502 replied:
Can you remove the Wrapper's code?

samip5 replied:
I did already though? @alextran1502

samip5 committed:
feat(server): Do not use a wrapper for ExifTool logging.

samip5 committed:
Merge branch 'main' into feat/exif

# Conflicts:
#	server/apps/microservices/src/processors/metadata-extraction.processor.ts

samip5 committed:
fix merge conflicts in metadata-extractor
